### ISSUE TITLE
Can't decode ill-formed UTF-8 octet sequence when using --set argument

### ISSUE THREAD

**Issue by joaostorrer** [sqitchers/sqitch#809]
Sqitch is failing whenever I set a variable with a value that contain accent:

```console
root@d975b89f1c63:/repo# sqitch deploy flipr_test --set var="ãçê"
Deploying changes to flipr_test
  + bla_heg ......... not ok
Can't decode ill-formed UTF-8 octet sequence <E3> at end of file at /bin/../lib/perl5/App/Sqitch.pm line 480.

Deploy failed
```

The same happens with `revert` and `rebase` commands.

I try set LC_ALL & LANG to pt_BR.UTF-8, but the error continues.

```console
root@d975b89f1c63:/repo# sqitch --version
sqitch (App::Sqitch) v1.4.0
```

```console
root@d975b89f1c63:/repo# perl -v

This is perl 5, version 32, subversion 1 (v5.32.1) built for x86_64-linux-gnu-thread-multi
(with 47 registered patches, see perl -V for more detail)

Copyright 1987-2021, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
```

**Comment by theory** [comment]
Hrm. That's [coming from the spool method](https://github.com/sqitchers/sqitch/blob/v1.4.0/lib/App/Sqitch.pm#L480), which spools output from the database CLI. Which RDBMS are you using? I wonder if it's not properly emitting UTF-8 for some reason.

**Comment by joaostorrer** [comment]
I'm using PostgreSQL.

**Comment by theory** [comment]
Is your database using UTF-8 encoding? Check [client_encoding](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-CLIENT-ENCODING).

**Comment by theory** [comment]
Sqitch requires everything to be UTF-8, and it sets up the Oracle connection for that:

https://github.com/sqitchers/sqitch/blob/7edc5366f1c801ac9f97f9ec0157e5fbdce39aac/lib/App/Sqitch/Engine/oracle.pm#L20-L27

So be sure that the contents of the variable are UTF-8 and not CP1252.

**Comment by joaostorrer** [comment]
I tried a few things to guarantee the variable is in UTF-8:

```console
root@18d3f3eb022a:/repo# sqitch deploy flipr_test --set var="$(echo "ãçê")"
Deploying changes to flipr_test
  + appschema ................. not ok
Can't decode ill-formed UTF-8 octet sequence <E3> at end of file at /bin/../lib/perl5/App/Sqitch.pm line 480.

Deploy failed
root@18d3f3eb022a:/repo# sqitch deploy flipr_test --set var="$(echo "ãçê" | iconv -t UTF-8)"
Deploying changes to flipr_test
  + appschema ................. not ok
Can't decode ill-formed UTF-8 octet sequence <E3> at end of file at /bin/../lib/perl5/App/Sqitch.pm line 480.

Deploy failed
root@18d3f3eb022a:/repo# echo "ãçê" | file -
/dev/stdin: UTF-8 Unicode text
root@18d3f3eb022a:/repo# echo -n "ãçê" | file -
/dev/stdin: UTF-8 Unicode text, with no line terminators
```

I also tested deploying in another Oracle Database that is configured with NLS_CHARACTERSET AL32UTF8, but I got the same error message.

**Comment by joaostorrer** [comment]
The patch works!

**Comment by theory** [comment]
Yes, the patch works because it's not decoding the output from SQL*Plus into UTF-8. Which means SQL*Plus is not emitting UTF-8. Clearly there's some additional configuration we need to make to get it to do so.

**Comment by theory** [comment]
Sure seems like setting `NSL_LANG` [should work](https://unix.stackexchange.com/a/318045).

**Comment by joaostorrer** [comment]
I tried setting `NLS_LANG` to different values like `BRAZILIAN PORTUGUESE_BRAZIL.UTF8`, `PORTUGUESE_BRAZIL.UTF8`, `AMERICAN_AMERICA.UTF8` and `AMERICAN_AMERICA.AL32UTF8`, but they all failed with `Can't decode ill-formed...`

**Comment by theory** [comment]
Yeah, Sqitch sets `NLS_LANG`, so anything you set it to doesn't work. Unless you tried setting it in the source?

https://github.com/sqitchers/sqitch/blob/7edc5366f1c801ac9f97f9ec0157e5fbdce39aac/lib/App/Sqitch/Engine/oracle.pm#L20-L27

**Comment by joaostorrer** [comment]
If I set to `BRAZILIAN PORTUGUESE_BRAZIL.UTF8` or `PORTUGUESE_BRAZIL.UTF8` in the source I get the same error.

**Comment by theory** [comment]
For some reason SQL*Plus is returning non-UTF-8 bytes. I do not understand why and don't have the resources to investigate. Maybe it's worth asking on Stack Overflow or an Oracle forum?

The key thing is to get SQL*Plus to emit only UTF-8 bytes. Though honestly I have no idea what it thinks it's emitting. Your change name is ASCII, so should be fine. When you tried the patch a few weeks ago, you said [it worked](#issuecomment-1912570419); what, exactly, did it emit? Very curious to know what it might be choking on.

**Comment by joaostorrer** [comment]
I inserted the content of `var` into a table to get what it emit.
The column was inserted with que upside down question mark (`¿`).
As far as I know, this only indicates that the database tried to inserted a character that is not recognized by SQL*Plus.

**Comment by theory** [comment]
Is the deploy script you wrote UTF-8 encoded?

**Comment by joaostorrer** [comment]
Yes, all the .sql scripts inside my project are UTF-8 (without BOM) encoded and Windows (CR LF).

**Comment by theory** [comment]
Did you ever sort this out, @joaostorrer? The mystery to me is why sqlplus emits non-UTF-8 bytes when Sqitch explicitly sets `NLS_LANG` to `AMERICAN_AMERICA.AL32UTF8`.
